### PR TITLE
fix(Core/Creature): Fix healer aggro

### PR DESCRIPTION
**Pull request by IceNet**
## CHANGES PROPOSED:
-  Fix healer aggro can pull at any distance

## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3655

## TESTS PERFORMED:
- Tested ingame on Debian 10

## HOW TO TEST THE CHANGES:
- Look at https://github.com/azerothcore/azerothcore-wotlk/issues/3655#issuecomment-723461965

## KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Review by Si1ker — APPROVED** [review]

**Comment by Racknar0** [comment]
Before PR

When you right click on a creature while it is healing you, you can pull it from any distance

image

After PR

Fixed the problem, you can no longer pull from range while healing

image

**Comment by Si1ker** [comment]
@IceNet  build failing 🐱

File: src/server/game/AI/CreatureAI.cpp [hunk 1]
```diff
@@ -114,15 +114,21 @@ void CreatureAI::MoveInLineOfSight(Unit* who)
 {
     if (me->GetVictim())
         return;
+	
+    if (me->GetCreatureType() == CREATURE_TYPE_NON_COMBAT_PET) // non-combat pets should just stand there and look good;)
+        return;
```
Comment (FrancescoBorzi) [inline src/server/game/AI/CreatureAI.cpp:119]: brackets please

File: src/server/game/AI/CreatureAI.cpp [hunk 2]
```diff
@@ -114,15 +114,21 @@ void CreatureAI::MoveInLineOfSight(Unit* who)
...
+
+    // Northshire Abbey - wolves fix. They can call assistance only outside the starting zone.
+    if (me->ToUnit()->getLevel() < 5 && me->GetAreaId() == 9)
+        return;
```
Comment (FrancescoBorzi) [inline src/server/game/AI/CreatureAI.cpp:123]: brackets please

File: src/server/game/Combat/ThreatManager.cpp [hunk 3]
```diff
@@ -124,18 +124,18 @@ void HostileReference::fireStatusChanged(ThreatRefStatusChangeEvent& threatRefSt
 
 void HostileReference::addThreat(float modThreat)
 {
+	if (!modThreat)
+        return;
```
Comment (FrancescoBorzi) [inline src/server/game/Combat/ThreatManager.cpp:128]: brackets please

File: src/server/game/Spells/Spell.cpp [hunk 4]
```diff
@@ -5273,13 +5270,17 @@ void Spell::HandleThreatSpells()
...
         // positive spells distribute threat among all units that are in combat with target, like healing
-        if (m_spellInfo->_IsPositiveSpell() && IsFriendly)
+        if (m_spellInfo->IsPositive())
             target->getHostileRefManager().threatAssist(m_caster, threatToAdd, m_spellInfo);
         // for negative spells threat gets distributed among affected targets
-        else if (!m_spellInfo->_IsPositiveSpell() && !IsFriendly && target->CanHaveThreatList())
+        else
+        {
+            if (!target->CanHaveThreatList())
+                continue;
```
Comment (FrancescoBorzi) [inline src/server/game/Spells/Spell.cpp:5280]: brackets please

File: src/server/game/AI/CreatureAI.cpp [hunk 5]
```diff
@@ -114,15 +114,21 @@ void CreatureAI::MoveInLineOfSight(Unit* who)
 {
     if (me->GetVictim())
         return;
+	
+    if (me->GetCreatureType() == CREATURE_TYPE_NON_COMBAT_PET) // non-combat pets should just stand there and look good;)
+        return;
+
+    // Northshire Abbey - wolves fix. They can call assistance only outside the starting zone.
+    if (me->ToUnit()->getLevel() < 5 && me->GetAreaId() == 9)
```
Comment (FrancescoBorzi) [inline src/server/game/AI/CreatureAI.cpp:122]: this looks a bit hacky

**Comment by FrancescoBorzi** [comment]
this contains quite a generic change and I feel like more testing is needed to make sure other mechanics still work fine

File: src/server/game/Combat/ThreatManager.cpp [hunk 6]
```diff
@@ -145,14 +145,7 @@ void HostileReference::addThreat(float modThreat)
...
 {
-    // Xinef: Do not allow to modify threat by percent if threat is negative (forced to big value < 0 by spells adding temporary threat)
-    // Xinef: When the temporary effect ends, temporary threat is added back which results in huge additional amount of threat
-    if (iThreat <= 0)
-        return;
-
-    float tmpThreat = iThreat;
-    AddPct(tmpThreat, percent);
-    addThreat(tmpThreat - iThreat);
+    addThreat(CalculatePct(iThreat, percent));
```
Comment (FrancescoBorzi) [inline src/server/game/Combat/ThreatManager.cpp:148]: why this change? just wondering

**Comment by IceNet** [comment]
Fix is not from me. I asked Silker if i should push it because i don't know if it's a hackfix or not
Do what you want with it

File: src/server/game/Spells/Spell.cpp [hunk 7]
```diff
@@ -2975,12 +2975,9 @@ SpellMissInfo Spell::DoSpellHitOnUnit(Unit* unit, uint32 effectMask, bool scaleA
             }
 
             // xinef: triggered spells should not prolong combat
-            if (unit->IsInCombat() && !m_spellInfo->HasAttribute(SPELL_ATTR3_NO_INITIAL_AGGRO) && !m_triggeredByAuraSpell)
```
Comment (Shard-MW) [inline src/server/game/Spells/Spell.cpp:2978]: Why do we remove `&& !m_triggeredByAuraSpell` ?

File: src/server/game/Combat/ThreatManager.cpp [hunk 8]
```diff
@@ -124,18 +124,18 @@ void HostileReference::fireStatusChanged(ThreatRefStatusChangeEvent& threatRefSt
...
-    if (modThreat != 0.0f)
-    {
-        ThreatRefStatusChangeEvent event(UEV_THREAT_REF_THREAT_CHANGE, this, modThreat);
-        fireStatusChanged(event);
-    }
+    ThreatRefStatusChangeEvent event(UEV_THREAT_REF_THREAT_CHANGE, this, modThreat);
+    fireStatusChanged(event);
 
-    if (isValid() && modThreat >= 0.0f)
+    if (isValid() && modThreat > 0.0f)
```
Comment (Goatform) [inline src/server/game/Combat/ThreatManager.cpp:138]: 👀

**Comment by BarbzYHOOL** [comment]
Guys, if you can help icenet, do it

He paid for the fixes and now he shares them (thx), he is not the developer so feel free to edit the code he sent!
There are unrelated changes in here. It's Icenet's first PR too, that's why

File: src/server/game/Combat/ThreatManager.cpp [hunk 9]
```diff
@@ -124,18 +124,20 @@ void HostileReference::fireStatusChanged(ThreatRefStatusChangeEvent& threatRefSt
 
 void HostileReference::addThreat(float modThreat)
 {
+	if (!modThreat)
```
Comment (Winfidonarleyan) [inline src/server/game/Combat/ThreatManager.cpp:127]: codestyle

**Comment by Si1ker** [comment]
@BarbzYHOOL  👏  +10
@IceNet  Thanks, totally appreciated! 🐱

**Comment by FrancescoBorzi** [comment]
@Si1ker  @IceNet I understand and wanted to underline that my comments were merely technical.

Thank you @IceNet for sharing this with us, I appreciate.

File: src/server/game/Combat/ThreatManager.cpp [hunk 10]
```diff
@@ -145,14 +147,7 @@ void HostileReference::addThreat(float modThreat)
 
 void HostileReference::addThreatPercent(int32 percent)
 {
-    // Xinef: Do not allow to modify threat by percent if threat is negative (forced to big value < 0 by spells adding temporary threat)
-    // Xinef: When the temporary effect ends, temporary threat is added back which results in huge additional amount of threat
-    if (iThreat <= 0)
-        return;
```
Comment (Viste) [inline src/server/game/Combat/ThreatManager.cpp:151]: Why?

File: src/server/game/Spells/Spell.cpp [hunk 11]
```diff
@@ -2975,12 +2975,9 @@ SpellMissInfo Spell::DoSpellHitOnUnit(Unit* unit, uint32 effectMask, bool scaleA
...
+            if (unit->IsInCombat() && !m_spellInfo->HasAttribute(SPELL_ATTR3_NO_INITIAL_AGGRO))
             {
-                // xinef: start combat with hostile unit...
-                if (Unit* hostile = unit->getAttackerForHelper())
-                    m_caster->CombatStart(hostile, true);
-                //m_caster->SetInCombatState(unit->GetCombatTimer() > 0, unit);
```
Comment (Viste) [inline src/server/game/Spells/Spell.cpp:2983]: why?

File: src/server/game/Spells/Spell.cpp [hunk 12]
```diff
@@ -5273,13 +5270,19 @@ void Spell::HandleThreatSpells()
...
             continue;
 
-        bool IsFriendly = m_caster->IsFriendlyTo(target);
         // positive spells distribute threat among all units that are in combat with target, like healing
-        if (m_spellInfo->_IsPositiveSpell() && IsFriendly)
+        if (m_spellInfo->IsPositive())
             target->getHostileRefManager().threatAssist(m_caster, threatToAdd, m_spellInfo);
         // for negative spells threat gets distributed among affected targets
-        else if (!m_spellInfo->_IsPositiveSpell() && !IsFriendly && target->CanHaveThreatList())
+        else
```
Comment (Viste) [inline src/server/game/Spells/Spell.cpp:5277]: why?

File: src/server/game/Combat/ThreatManager.cpp (hunk 10 above)
Comment (Winfidonarleyan) [reply src/server/game/Combat/ThreatManager.cpp:151]: why not? ;D
